synapse/handlers/e2e_keys.py

```diff
@@ -497,6 +500,19 @@ async def query_local_devices(
             local_query, include_displaynames
         )
 
+        # Check if the application services have any additional results.
+        if self._query_appservices_for_keys:
+            # Query the appservices for any keys.
+            appservice_results = await self.appservice_handler.query_keys(query)
+
+            # Merge results, overriding with what the appservice returned.
+            for user_id, devices in appservice_results.get("device_keys", {}).items():
+                # Copy the appservice device info over the homeserver device info, but
+                # don't completely overwrite it.
+                results.setdefault(user_id, {}).update(devices)
+
+            # TODO Handle cross-signing keys.
```

Comment:
This is a pretty big TODO compared to what the MSC says, but it isn't abundantly clear how to merge the cross-signing info (`master_keys` / `self_signing_keys` / `user_signing_keys`).

Reply:
Also -- that will likely require some refactoring since `get_cross_signing_keys_from_cache` gets called after this.

Reply:
The MSC now kind of shrugs at this 🤷 , see https://github.com/matrix-org/matrix-spec-proposals/pull/3984#discussion_r1148109681

I think it is probably OK to gloss over this for now.